framework/src/processing/lib/spark-runtime/emr-containers/emr-eks-cluster-helpers.ts

```diff
@@ -0,0 +1,589 @@
+  karpenterAccount.addToPrincipalPolicy(karpenterControllerPolicyStatementIAM);
+  karpenterAccount.addToPrincipalPolicy(allowScopedEC2InstanceActions);
+  karpenterAccount.addToPrincipalPolicy(allowScopedEC2InstanceActionsWithTags);
+  karpenterAccount.addToPrincipalPolicy(allowScopedEC2LaunchTemplateActions);
+  karpenterAccount.addToPrincipalPolicy(allowMachineMigrationTagging);
+  karpenterAccount.addToPrincipalPolicy(allowScopedResourceCreationTagging);
+  karpenterAccount.addToPrincipalPolicy(allowScopedDeletion);
+  karpenterAccount.addToPrincipalPolicy(allowInterruptionQueueActions);
+  karpenterAccount.addToPrincipalPolicy(allowAPIServerEndpointDiscovery);
+  karpenterAccount.addToPrincipalPolicy(allowInstanceProfileReadActions);
+
+  //Deploy Karpenter Chart
+  const karpenterChart = cluster.addHelmChart('KarpenterHelmChart', {
+    chart: 'karpenter',
+    release: 'karpenter',
+    repository: 'oci://public.ecr.aws/karpenter/karpenter',
+    namespace: 'karpenter',
+    version: karpenterVersion,
+    timeout: Duration.minutes(14),
+    wait: true,
+    values: {
+      serviceAccount: {
+        name: 'karpenter',
+        create: false,
+        annotations: {
+          'eks.amazonaws.com/role-arn': karpenterAccount.role.roleArn,
+        },
+      },
+      settings: {
+        aws: {
+          defaultInstanceProfile: instanceProfile.instanceProfileName,
+          clusterName: eksClusterName,
+          clusterEndpoint: cluster.clusterEndpoint,
+          interruptionQueueName: karpenterInterruptionQueue.queueName,
+        },
+      },
+
+    },
+  });
+
+  karpenterChart.node.addDependency(karpenterAccount);
+
+  const karpenterInstancesSg = new SecurityGroup(scope, 'karpenterSg', {
+    vpc: cluster.vpc,
+    allowAllOutbound: true,
+    description: 'security group for a karpenter instances',
+    securityGroupName: 'karpenterSg',
+    disableInlineRules: true,
+  });
+
+  Tags.of(karpenterInstancesSg).add('karpenter.sh/discovery', `${eksClusterName}`);
+
+  cluster.clusterSecurityGroup.addIngressRule(
+    karpenterInstancesSg,
+    Port.allTraffic(),
+  );
+
+  karpenterInstancesSg.addIngressRule(
+    karpenterInstancesSg,
+    Port.allTraffic(),
+  );
+
+  karpenterInstancesSg.addIngressRule(
+    cluster.clusterSecurityGroup,
+    Port.allTraffic(),
+  );
+
+  Tags.of(cluster.vpc).add(
+    'karpenter.sh/discovery', eksClusterName,
+  );
+
+  cluster.vpc.privateSubnets.forEach((subnet) => {
+    Tags.of(subnet).add('karpenter.sh/discovery', eksClusterName);
+  });
+
+  cluster.vpc.publicSubnets.forEach((subnet) =>
+    Tags.of(subnet).add('karpenter.sh/discovery', eksClusterName),
+  );
+
+  const privateSubnets = cluster.vpc.selectSubnets({
+    onePerAz: true,
+    subnetType: SubnetType.PRIVATE_WITH_EGRESS,
+  }).subnets;
+
+  let manifest = Utils.readYamlDocument(`${__dirname}/resources/k8s/karpenter-provisioner-config/${karpenterVersion}/tooling-provisioner.yml`);
```

Comment:
everything starting from here should be isolated in a dedicated method because it only concerns the tooling provisioner